## Proyecto de Análisis de Datos con Python - Diploma UCU - Nicolás Gallo

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

# Subimos un nivel desde notebooks/ hacia la raíz del proyecto
base_dir = os.path.abspath(os.path.join(os.getcwd(), '..'))
ruta = os.path.join(base_dir, 'data', 'raw', 'p4ds_grupo_c_egresos_hospitalarios.csv')

print("Buscando en:", ruta)

df = pd.read_csv(ruta, sep=";")

display(df.head())
print(f"Filas: {df.shape[0]:,} | Columnas: {df.shape[1]}")
df.info()
print("\n=== VALORES NULOS ===")
print(df.isnull().sum())

Buscando en: C:\Users\usuario\PycharmProjects\nicolas_gallo_diploma_proyecto_python\data\raw\p4ds_grupo_c_egresos_hospitalarios.csv


,ID,AÑO,SECTOR,GENERO,GRUPO ETAREO,REGION,DIAGNOSTICO,CAUSA EXTERNA
0,1,2024,PRIVADO,F,65+,INTERIOR,Hemorragia intracraneal,NC
1,2,2024,PRIVADO,M,65+,INTERIOR,Tumor maligno de la vejiga urinaria,NC
2,3,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades del sistema urinario,NC
3,4,2024,PRIVADO,F,65+,INTERIOR,Otras enfermedades del sistema urinario,NC
4,5,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades del sistema urinario,NC


Filas: 1,132,011 | Columnas: 8
<class 'pandas.DataFrame'>
RangeIndex: 1132011 entries, 0 to 1132010
Data columns (total 8 columns):
 #   Column         Non-Null Count    Dtype
---  ------         --------------    -----
 0   ID             1132011 non-null  int64
 1   AÑO            1132011 non-null  int64
 2   SECTOR         1132011 non-null  str  
 3   GENERO         1132011 non-null  str  
 4   GRUPO ETAREO   1132011 non-null  str  
 5   REGION         1089284 non-null  str  
 6   DIAGNOSTICO    1132011 non-null  str  
 7   CAUSA EXTERNA  1132011 non-null  str  
dtypes: int64(2), str(6)
memory usage: 144.4 MB

=== VALORES NULOS ===
ID                   0
AÑO                  0
SECTOR               0
GENERO               0
GRUPO ETAREO         0
REGION           42727
DIAGNOSTICO          0
CAUSA EXTERNA        0
dtype: int64


In [3]:
print("=== VALORES ÚNICOS POR COLUMNA ===")
for col in df.columns:
    print(f"\n{col} ({df[col].nunique()} únicos):")
    print(df[col].unique()[:10])  # primeros 10 valores únicos

=== VALORES ÚNICOS POR COLUMNA ===

ID (1048158 únicos):
[ 1  2  3  4  5  6  7  8  9 10]

AÑO (4 únicos):
[2024 2023 2022 2021]

SECTOR (2 únicos):
<ArrowStringArray>
['PRIVADO', 'PUBLICO']
Length: 2, dtype: str

GENERO (2 únicos):
<ArrowStringArray>
['F', 'M']
Length: 2, dtype: str

GRUPO ETAREO (10 únicos):
<ArrowStringArray>
[      '65+',   '45 a 64',   '25 a 44',   '20 a 24',   '15 a 19',   '10 a 14',
     '5 a 9',     '1 a 4', 'Menor a 1', 'Sin datos']
Length: 10, dtype: str

REGION (4 únicos):
<ArrowStringArray>
['INTERIOR', 'MONTEVIDEO', nan, 'EXTERIOR', 'Sin datos']
Length: 5, dtype: str

DIAGNOSTICO (258 únicos):
<ArrowStringArray>
[                     'Hemorragia intracraneal',
          'Tumor maligno de la vejiga urinaria',
      'Otras enfermedades del sistema urinario',
                           'Fractura del fémur',
                     'Traumatismo intracraneal',
                          'Insuficiencia renal',
  'Otras enfermedades del sistema respiratorio',
 'Otros 

In [5]:
print("== DUPLICADOS ===")
print(f"Filas duplicadas: {df.duplicated().sum():,}")
print(f"IDs duplicados: {df['ID'].duplicated().sum():,}")

== DUPLICADOS ===
Filas duplicadas: 83,853
IDs duplicados: 83,853


In [6]:
# Mostramos algunos ejemplos de filas duplicadas
duplicados = df[df.duplicated(keep=False)]
print(f"Total de filas involucradas en duplicados: {len(duplicados):,}")
display(duplicados.sort_values('ID').head(10))

Total de filas involucradas en duplicados: 167,706


,ID,AÑO,SECTOR,GENERO,GRUPO ETAREO,REGION,DIAGNOSTICO,CAUSA EXTERNA
0,1,2024,PRIVADO,F,65+,INTERIOR,Hemorragia intracraneal,NC
1127284,1,2024,PRIVADO,F,65+,INTERIOR,Hemorragia intracraneal,NC
14,15,2024,PRIVADO,M,65+,INTERIOR,"Otros traumatismos de regiones especificadas, ...",Todas las demás causas externas
1116603,15,2024,PRIVADO,M,65+,INTERIOR,"Otros traumatismos de regiones especificadas, ...",Todas las demás causas externas
1094611,32,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades isquémicas del corazón,NC
31,32,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades isquémicas del corazón,NC
1114159,52,2024,PRIVADO,M,65+,INTERIOR,Fracturas de otros huesos de los miembros,Accidentes de transporte
51,52,2024,PRIVADO,M,65+,INTERIOR,Fracturas de otros huesos de los miembros,Accidentes de transporte
1068065,53,2024,PRIVADO,M,65+,INTERIOR,"Tumor maligno de la tráquea, de los bronquios ...",NC
52,53,2024,PRIVADO,M,65+,INTERIOR,"Tumor maligno de la tráquea, de los bronquios ...",NC


In [7]:
# Eliminamos duplicados conservando la primera aparición
df_limpio = df.drop_duplicates(keep='first')

print(f"Filas originales:  {len(df):,}")
print(f"Filas después:     {len(df_limpio):,}")
print(f"Filas eliminadas:  {len(df) - len(df_limpio):,}")

Filas originales:  1,132,011
Filas después:     1,048,158
Filas eliminadas:  83,853


In [8]:
# Veamos cuántos registros tienen valores problemáticos
print("=== VALORES A LIMPIAR ===")

print(f"\nREGION - nulos: {df_limpio['REGION'].isna().sum():,}")
print(f"REGION - 'Sin datos': {(df_limpio['REGION'] == 'Sin datos').sum():,}")

print(f"\nGRUPO ETAREO - 'Sin datos': {(df_limpio['GRUPO ETAREO'] == 'Sin datos').sum():,}")

print(f"\nCAUSA EXTERNA - '99': {(df_limpio['CAUSA EXTERNA'] == '99').sum():,}")
print(f"CAUSA EXTERNA - 'SD': {(df_limpio['CAUSA EXTERNA'] == 'SD').sum():,}")
print(f"CAUSA EXTERNA - 'NC': {(df_limpio['CAUSA EXTERNA'] == 'NC').sum():,}")

=== VALORES A LIMPIAR ===

REGION - nulos: 39,597
REGION - 'Sin datos': 2,521

GRUPO ETAREO - 'Sin datos': 2

CAUSA EXTERNA - '99': 11,192
CAUSA EXTERNA - 'SD': 1,769
CAUSA EXTERNA - 'NC': 947,609


##Muy bien, ahora tenemos el panorama completo. Analicemos cada caso antes de limpiar:
REGION ⚠️

39,597 nulos + 2,521 "Sin datos" = 42,118 registros sin región (~4% del total) — los unificamos como 'Sin datos'

GRUPO ETAREO ✅

Solo 2 registros con "Sin datos" — insignificante, los dejamos así

CAUSA EXTERNA 🔍

'NC' = 947,609 registros (~90% del total) — significa "No Corresponde", es decir egresos sin causa externa. Es un valor válido, no un error
'SD' = 1,769 → Sin Dato
'99' = 11,192 → también parece código para Sin Dato

Los unificamos: 'SD' y '99' los convertimos a 'Sin datos', y 'NC' lo dejamos como está.

In [9]:
# Unificamos REGION: nulos y 'Sin datos' → 'Sin datos'
df_limpio['REGION'] = df_limpio['REGION'].fillna('Sin datos')

# Unificamos CAUSA EXTERNA: '99' y 'SD' → 'Sin datos'
df_limpio['CAUSA EXTERNA'] = df_limpio['CAUSA EXTERNA'].replace({'99': 'Sin datos', 'SD': 'Sin datos'})

print("=== VERIFICACIÓN POST LIMPIEZA ===")
print("\nREGION:")
print(df_limpio['REGION'].value_counts())
print("\nCAUSA EXTERNA:")
print(df_limpio['CAUSA EXTERNA'].value_counts())

=== VERIFICACIÓN POST LIMPIEZA ===

REGION:
REGION
INTERIOR      620930
MONTEVIDEO    384841
Sin datos      42118
EXTERIOR         269
Name: count, dtype: int64

CAUSA EXTERNA:
CAUSA EXTERNA
NC                                                                   947609
Caídas                                                                32517
Todas las demás causas externas                                       27063
Accidentes de transporte                                              16654
Sin datos                                                             12961
Lesiones autoinflingidas intencionalmente                              7079
Agresiones                                                             3783
Envenenamiento accidental por, y exposición a, sustancias nocivas       368
Exposición al humo, fuego y llamas                                      123
Ahogamiento y sumersión accidentales                                      1
Name: count, dtype: int64


In [10]:
#El dataset está limpio. Ahora hacemos un resumen final antes de grabarlo:
print("=== RESUMEN FINAL DEL DATASET LIMPIO ===")
print(f"\nFilas originales:  1,132,011")
print(f"Filas finales:     {len(df_limpio):,}")
print(f"Columnas:          {df_limpio.shape[1]}")
print(f"\nValores nulos restantes:")
print(df_limpio.isnull().sum())

=== RESUMEN FINAL DEL DATASET LIMPIO ===

Filas originales:  1,132,011
Filas finales:     1,048,158
Columnas:          8

Valores nulos restantes:
ID               0
AÑO              0
SECTOR           0
GENERO           0
GRUPO ETAREO     0
REGION           0
DIAGNOSTICO      0
CAUSA EXTERNA    0
dtype: int64


In [11]:
#Ahora grabamos el CSV limpio en data/processed/:
# Grabamos el dataset limpio
ruta_salida = os.path.join(base_dir, 'data', 'processed', 'egresos_hospitalarios_limpio.csv')

df_limpio.to_csv(ruta_salida, index=False, sep=';')

print(f"Archivo grabado en: {ruta_salida}")
print(f"Filas grabadas: {len(df_limpio):,}")

Archivo grabado en: C:\Users\usuario\PycharmProjects\nicolas_gallo_diploma_proyecto_python\data\processed\egresos_hospitalarios_limpio.csv
Filas grabadas: 1,048,158


In [ ]:
#La Fase 1 está completa:

#Cargamos el dataset (1,132,011 filas)
#Eliminamos 83,853 filas duplicadas
#Unificamos nulos y 'Sin datos' en REGION
#Unificamos '99' y 'SD' en CAUSA EXTERNA
#Grabamos el CSV limpio en data/processed/